gazebo_plugins/test/test_gazebo_ros_wheelslip.cpp

```diff
@@ -258,6 +263,57 @@ TEST_F(GazeboRosWheelSlipTest, TestSetParameters)
   }
 }
 
+class GazeboRosWheelSlipPublisherTest : public gazebo::ServerFixture
+{
+};
+
+TEST_F(GazeboRosWheelSlipPublisherTest, Publishing)
+{
+  // Load test world and start paused
+  this->Load("worlds/gazebo_ros_wheel_slip.world", true);
+
+  // World
+  auto world = gazebo::physics::get_world();
+  ASSERT_NE(nullptr, world);
+
+  // Create node and executor
+  auto node = std::make_shared<rclcpp::Node>("gazebo_ros_joint_state_publisher_test");
+  ASSERT_NE(nullptr, node);
+
+  rclcpp::executors::SingleThreadedExecutor executor;
+  executor.add_node(node);
+
+  // Create subscriber
+  gazebo_msgs::msg::WheelSlip::SharedPtr latestMsg;
+  auto sub = node->create_subscription<gazebo_msgs::msg::WheelSlip>(
+    "trisphere_cycle_slip/wheel_slip", rclcpp::QoS(1),
+    [&latestMsg](const gazebo_msgs::msg::WheelSlip::SharedPtr msg) {
+      latestMsg = msg;
+    });
+
+  // Spin until we get a message or timeout
+  auto startTime = std::chrono::steady_clock::now();
+  while (latestMsg == nullptr &&
+    (std::chrono::steady_clock::now() - startTime) < std::chrono::seconds(2))
+  {
+    world->Step(100);
+    executor.spin_once(100ms);
+    gazebo::common::Time::MSleep(100);
+  }
+
+  // Check that we receive the latest joint state
+  ASSERT_NE(nullptr, latestMsg);
+
+  ASSERT_LT(0u, latestMsg->name.size());
+  ASSERT_GT(4u, latestMsg->name.size());
+  ASSERT_EQ(latestMsg->name.size(), latestMsg->lateral_slip.size());
+  ASSERT_EQ(latestMsg->name.size(), latestMsg->longitudinal_slip.size());
+
+  EXPECT_NEAR(0.0, latestMsg->lateral_slip[0], tol);
+  ASSERT_LT(0.0, fabs(latestMsg->longitudinal_slip[0]));
+  ASSERT_GT(0.7, fabs(latestMsg->longitudinal_slip[0]));
```

Comment:
given our discussion of the sign convention for longitudinal slip, I recommend an expectation that doesn't use absolute value here. And I think we expect it to be positive for this test? If possible, `EXPECT_NEAR(0.3 /* or whatever */, latestMsg->longitudinal_slip[0], generous_tolerance);` would be nice.

side note: failing an `ASSERT_*` will immediately exit the test, so I only use that for things like checking for null pointers that I'm about to dereference. I recommend `EXPECT_*` for most other expectations

Reply:
In the simulation, I believe the tricycle's front wheel gives a different sign for the longitudinal wheel slip than the rear wheels. This is why I use the absolute value. I assume this has to be fixed in the tricycle model, or maybe I can add another parameter for flipping the sign of the calculated slip.

The wheel slip publisher publishes information for a random set of the wheels with the plugin. Alternatively, I could check for the name of each wheel and then test its slip with `EXPECT_NEAR`, if you think this would be better.

Reply:
the controller is only applying torque to the rear wheels, so that's why the slip values are different:

https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1331/files#diff-fd5ababb339b088418f3745d7340ba6b7cee97a16c5d34d9e8abbe3aa72e1461R48-R58

if you want to get a single wheel-slip message with all three wheels, you can unify the `gazebo_ros_wheel_slip` plugin instances (since there's currently two plugins). Then you could check the wheel name and set expectations accordingly.

Or, you can enable a velocity controller for the front wheel as well. I think there's a few options.

Reply:
Makes sense, thank you.